### ISSUE TITLE
Ractive.js v0.8.0-edge simple update problem

### ISSUE THREAD

**Issue by Jeff17Robbins** [ractivejs/ractive#2518]
Ractive.js v0.8.0-edge still has an update() problem

	Ractive.js v0.8.0-edge
	Wed Apr 20 2016 23:39:55 GMT+0000 (UTC) - commit abdad6a6cdfe8c66d4900c603b0417389999862f

http://jsfiddle.net/xr4vzk77/ (Ractive.js v0.7.3) works

http://jsfiddle.net/8x4bq41b/ (Ractive.js v0.8.0-edge) doesn't work

HTML
```html
  <div id="output"></div>  
  <script id="template" type="text/ractive">
    <b>{{status}}</b>
  </script>
```

JavaScript
```javascript
var r = new Ractive({
  el: '#output',
  template: '#template'
});

r.on('change', function(updates) {
  console.log(updates);
});

r.set('status', 'bad');
var model = r.get();
model.status = 'OK';
r.update('status');
```

Should see **OK** on the page and in the console

**Comment by martypdx** [comment]
The issue is that we added the return of a synthetic object in [root model](https://github.com/ractivejs/ractive/blob/dev/src/model/RootModel.js#L65) `.get`'s in order to include the mapped and computed properties.

It doesn't happen when you use a named key instead:

```js
var model = r.get( 'model' );
model.status = 'OK';
r.update('model.status');
```

Adding an options parameter to `.get` seems way too broad. If you don't need to access mapped keys (those passed in or implicitly resolved from the parent) we could distinguish between:
* `ractive.get()` - get root data plus mapped and props, and 
* `ractive.get('.')` - get the root data object of this instance (mostly equiv template, except no mapped)

Thoughts?

**Comment by evs-chris** [comment]
I lean towards an options map as a second parameter, but `'.'` seems reasonable. It's a bit magical, but sometimes magic is good :smile: and it's consistent with other established conventions.

**Comment by Jeff17Robbins** [comment]
I think I understand. 

In Ractive.js v0.7.3, ractive.get() returns the actual model, without computeds.

http://jsfiddle.net/g60h62nv/ (Ractive.js v0.7.3) logs this to the console:

`Object {status: "OK"}`

Whereas http://jsfiddle.net/3uc0g182/ (Ractive.js v0.8.0-edge)  logs this to the console:

` Object  status: "OK"  superStatus: "OKOK"  `


It always bothered me that the computeds were, in effect, hidden in Ractive.js v0.7.3, so I like this change!


@martypdx what are "mapped" properties?  You refer to them also, but I don't know what that means.  I also don't understand "If you don't need to access mapped keys (those passed in or implicitly resolved from the parent)"    What are "mapped keys", and how do they differ from "mapped properties"?   

Would `ractive.get('.')` be Ractive.js v0.8.0-edge's equivalent to `ractive.get()` in Ractive.js v0.7.3?

**Comment by martypdx** [comment]
> I lean towards an options map as a second parameter

@evs-chris normally I would too, but for this use it means for majority of cases like `ractive.get( 'model', { someOption: true } );`, it doesn't actually do anything. Whereas using the blank vs `.` syntax localizes it to the actual issue of root gets.

> @martypdx what are "mapped" properties? You refer to them also, but I don't know what that means. I also don't understand "If you don't need to access mapped keys (those passed in or implicitly resolved from the parent)" 

@Jeff17Robbins data properties passed from the parent are "mapped":

```html
<some-component foo='{{bar}}'/>
```

Meaning that the data `foo` does not "live" on the `some-component` instance data root, all calls are "mapped" to the `bar` data on the parent instance data root:

```js
var ractive = new Ractive({
    template: `<some-component foo='{{bar}}'/>`,
    data: {
        bar: 'Le Bar'
    },
    components: {
        'some-component': Ractive.extend({ 
            template: `{{foo}} {{qux}}`,
            data: { qux: 'quxify' }
        })
    }
})

ractive.get();
// { bar: 'Le Bar' }
var component = ractive.findComponent('some-component');
component.get();
// { foo: 'LeBar', qux: 'quxify' }
component.get('.') //proposed, only the actual root model
// { qux: 'quxify' }

```
> What are "mapped keys", and how do they differ from "mapped properties"?

Just me using two different words, sorry.

> Would ractive.get('.') be Ractive.js v0.8.0-edge's equivalent to ractive.get() in Ractive.js v0.7.3?

Yes

> Would it be ok to keep ractive.get() working the same way as in Ractive.js v0.7.3 unless specified by an options parameter? Or do new things inside Ractive.js v0.8.0-edge depend on this new behavior?

While it is a breaking change, IMO it's a better API for reasons listed above.

**Comment by Jeff17Robbins** [comment]
Further thoughts:

There are 3 Javascript ways to access nominally "all" of a Ractive instance's data:

1. `ractive.get()`
2. `ractive.observe('*')`
3. `ractive.on('change')`

It would be great to align all 3 ways to make things easier to understand.  

For example, ractive.observe takes an `init` option that 

> Defaults to true. Whether or not to initialise the observer, i.e. call the function with the current value of keypath as the first argument and undefined as the second 

> In the case of computeds, ractive.observe's `initialise` callback provides an evaluated value of each computed.(http://docs.ractivejs.org/latest/ractive-observe)

Yet, `ractive.get()` does not provide an evaluated value for each computed, but, rather, provides `undefined` as the value of each computed.  Unless...unless the computed is used in a mustache in the ractive's template, in which case the render logic does an explicit ractive.get('computed') which **does** force the evaluation of the computed.

Some of this illustrated here for edge http://jsfiddle.net/ezhLscu9/
(and here for 0.7.3 http://jsfiddle.net/wexgmv1f/)

It would be nice to have a choice as to whether or not the computeds get evaluated.

One idea would be to support `ractive.get('*')` and have this return the same thing all at once that the successive callbacks to `ractive.observe('*')` would have yielded key by key in the default `initialise` run-through.  This would allow code consistent access to all the data, whether "pulled" via `ractive.get()` or "pushed" via `ractive.observe()`.  

In other words, the current Ractive.js v0.8.0-edge behavior for `ractive.get()` gives you results for your computeds that depends on whether or not a given computed is used in your template.  Whereas `ractive.observe('*')` forces evaluation of all computeds, as if they all appeared in mustaches.  The proposed `ractive.get('*')` would get you all your data and all your computeds evaluated.  

[I can't speak usefully about the other mapped properties used in the Ractive components feature, simply because we aren't using the components feature and so I have no experience with it.]

**Comment by martypdx** [comment]
> Unless...unless the computed is used in a mustache in the ractive's template, in which case the render logic does an explicit ractive.get('computed') which does force the evaluation of the computed.

@Jeff17Robbins  That's a bug! 

Computed should return consistently if included on get. If you do `r.get('indy')` you get the correct value. 

When I added the get for computed, I didn't consider they might not be evaluated yet. 😊  

I think it means changing [this line](https://github.com/ractivejs/ractive/blob/dev/src/model/RootModel.js#L68) to use `.get()` instead of `.value`

**Comment by Jeff17Robbins** [comment]
Great (not about the bug, but about the intent!)  So `ractive.get()` will get me "everything".   

I'm still slogging away with our app (which runs nicely on Ractive.js v0.7.3) to identify Ractive-edge-related problems.  There are at least a couple more (involving .update() and/or the change event, and one involving the Array adaptor.)  I'm trying to create smallest reproductions of these bugs.  More later.

**Comment by martypdx** [comment]
> I'm still slogging away with our app (which runs nicely on Ractive.js v0.7.3) to identify Ractive-edge-related problems. There are at least a couple more (involving .update() and/or the change event, and one involving the Array adaptor.) I'm trying to create smallest reproductions of these bugs. More later.

@Jeff17Robbins thanks for your help!

**Comment by lazylester** [comment]
late to the discussion here, but having just discovered this enhancement my comment is this...

Given several possible implementations to distinguish the scope of the get() method, my preference would be to choose one of the non-breaking approaches.

So expanding get() to fetch all (including computed) attributes, while get('.') only fetches non-computed attributes seems backwards to me. Also as @evs-chris mentions, '.' is a little magical and (for me) is not intention-revealing.

My suggestion would be to leave get() retrieving only non-computed attributes, perhaps in addition to get({computed:false}), and get({computed: true}) as the implementation of the new feature.

**Comment by guilhermeaiolfi** [comment]
But how does it affect your app if ```.get()``` returns more than what you're expecting? And why you make the difference between mapped properties and computed ones? Isn't it easier to get all your instance properties and just ignore what you don't want?

I'm failing to see how that's a problem. I'm would like to get all properties (mapped and computed) in a simple ```.get()```

**Comment by lazylester** [comment]
@guilhermeaiolfi well firstly there's the "philosophical" issue that non-breaking changes should be preferred over breaking changes, all other things being equal, in my opinion.

But why is it a breaking change to now return computed properties in addition to the mapped properties, with the same method that previously returned only mapped properties? Well I ran into the problem in a scenario where I have an "edit-in-place" behaviour on the page. So before switching to edit mode, I make a stashed copy of the object that I can restore if editing is cancelled. I make the stashed copy with all the attributes returned by get(), something like this: 
`this.stashed_copy = $.extend(true,{},this.get())`

 And I restore it with:
`this.set(this.stashed_copy)`

It works well in 0.7.3, but the restore fails if read-only computed attributes are included in get().

That's my thinking. Thankfully my tests caught this, but I'm always worried that a library could introduce a breaking change that my tests don't catch. Perhaps my scenario is obscure and nobody else does this kind of thing, but this is what prompts me to express my preference for a non-breaking change. There might be other scenarios that I haven't thought about where a change in the behaviour would break someone's app.

**Comment by lazylester** [comment]
@guilhermeaiolfi Another scenario where returning computed attributes would be a breaking change is when an instance is persisted to a server. The instance may have computed properties that relate only to its rendering and which would not normally be persisted to the server database. If now the set of properties returned by get() is expanded, the back-end will probably complain (e.g. Rails strong parameters, or no database columns for the computed properties).

**Comment by evs-chris** [comment]
edge now has an option on get to return the actual object or a virtual object composed of the data, any links/mappings, and computations. For root e.g. `ractive.get()` it defaults to true, but if you want just the data, you can used `ractive.get({ virtual: false })`. For nested paths, it defaults to false, but if you want to include links (and maybe computations if we add that at some point?), you can use `ractive.get('some.path', { virtual: true })`.
